File: Golang_src/easy/numberComplement_476.go
package easy

// Use Bit Manipulation:
// Time Complexity: O(1)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 100.00%
func findComplement(num int) int {
	mask := 1 << 30

	for num&mask == 0 {
		mask >>= 1
	}

	return num ^ (mask<<1 - 1)
}

File: Cpp_src/easy/number_complement_476.cc
#include <iostream>
using namespace std;

// Use bit manipulation with iteration
// Time Complexity: O(1)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 6.2 MB, less than 100.00%
class Solution
{
public:
    int findComplement(int num)
    {
        unsigned int mask = ~0;
        while (mask & num)
        {
            mask <<= 1;
        }

        return num ^ ~mask;
    }
};